nltk/parse/corenlp.py

```diff
@@ -0,0 +1,649 @@
+    sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+    sock.bind(('', port))
+
+    p = sock.getsockname()[1]
+    sock.close()
+
+    return p
+
+
+class CoreNLPServer(object):
+
+    _MODEL_JAR_PATTERN = r'stanford-corenlp-(\d+)\.(\d+)\.(\d+)-models\.jar'
+    _JAR = r'stanford-corenlp-(\d+)\.(\d+)\.(\d+)\.jar'
+
+    def __init__(
+        self, path_to_jar=None, path_to_models_jar=None, verbose=False,
+        java_options=None, corenlp_options=None, port=None,
+    ):
+
+        jars = list(find_jar_iter(
+            self._JAR,
+            path_to_jar,
+            env_vars=('CORENLP', ),
+            searchpath=(),
+            url=_stanford_url,
+            verbose=verbose,
+            is_regex=True,
+        ))
+
+        # find the most recent code and model jar
+        stanford_jar = max(
+            jars,
+            key=lambda model_name: re.match(self._JAR, model_name)
+        )
+
+        if port is None:
+            try:
+                port = try_port(9000)
+            except socket.error:
+                port = try_port()
+
+            if corenlp_options:
+                corenlp_options.append(str(port))
+            else:
+                corenlp_options = [str(port)]
+        else:
+            try_port(port)
+
+        self.url = 'http://localhost:{}'.format(port)
+
+        model_jar = max(
+            find_jar_iter(
+                self._MODEL_JAR_PATTERN,
+                path_to_models_jar,
+                env_vars=('CORENLP_MODELS', ),
+                searchpath=(),
+                url=_stanford_url,
+                verbose=verbose,
+                is_regex=True,
+            ),
+            key=lambda model_name: re.match(self._MODEL_JAR_PATTERN, model_name)
+        )
+
+        self.verbose = verbose
+
+        self._classpath = stanford_jar, model_jar
+
+        self.corenlp_options = corenlp_options or []
+        self.java_options = java_options or ['-mx2g']
+
+    def start(self):
+        cmd = ['edu.stanford.nlp.pipeline.StanfordCoreNLPServer']
+
+        if self.corenlp_options:
+            cmd.extend(self.corenlp_options)
+
+        # Configure java.
+        default_options = ' '.join(_java_options)
+        config_java(options=self.java_options, verbose=self.verbose)
+
+        try:
+            # TODO: it's probably a bad idea to pipe stdout, as it will
+            #       accumulate when lots of text is being parsed.
+            self.popen = java(
+                cmd,
+                classpath=self._classpath,
+                blocking=False,
+                stdout='pipe',
+                stderr='pipe',
+            )
+        finally:
+            # Return java configurations to their default values.
+            config_java(options=default_options, verbose=self.verbose)
+
+        # Check that the server is istill running.
+        # TODO: is there a better way of checking whether a server is ready to
+        #       accept connections?
+        time.sleep(5)
```

Comment:
Yes there is! You can wait for a 2XX response on the `/ready` endpoint, which should happen once all the setup tasks are complete. The `/live` endpoint should tell you if the server booted at all. This conforms to the Kubernetes liveness and readiness probes.